### ISSUE TITLE
Zeebe Configuration Milestone Summary

### ISSUE THREAD

**Issue by pihme** [zeebe-io/zeebe#3824]
## Goal
Users should be able to configure Zeebe Broker and Gateway through environment variables and YAML configuration files in an easy and consistent manner.

## Requirements
* Should allow reading config from file or environment variables (dev, ops)
* Should allow overlays of different configuration files/sources (dev, ops)
  - environment variables
  - YAML files
  - settings in environment variables shall overwrite settings in files
* Should use standard file format (dev, ops)
  - That can be easily generated in different programming languages (dev, ops)

## Scope
* Configuration for broker and gateway

### Out of Scope
* Changing content of configuration / breaking up configuration into smaller chunks
* Reload of configuration changes at runtime
* Distribution of configuration between brokers / consistency checks between configurations of different brokers
* Config sources like databases, Zookeeper etc.

### Stakeholders
* Dev - "Should be easy to use and more mainstream then the current TOML config files" 
* Ops - "Basically, we want to make all important settings with environment variables"

### Status Quo
* Configuration can be read from environment variables or TOML file
* Configuration validation is rather weak and incomplete
* Configuration read once at startup
* Configuration is read only
* Cluster assumes that all brokers are configured identically or at least compatibly
* Changes to configuration require a restart to take effect
* Not all settings are known ahead of time; e.g. nodeID and initialContactPoints will only be known and startup

### Pain Pints with Status Quo
* Config settings not really visible in testing/difficult to test (dev)
* Unit tests primarily focus on correct parsing of configuration (dev)
* DurationUtil, ByteValue used throughout the code base, not just in the parsing layer (dev)
* Only single file can be read, no overlay possible (dev, ops)
* Complex configuration is hard to put into env variables (dev, ops)
* TOML format not supported/hard to write across different programming languages (dev, ops)
* Not all settings are available as environment variables (ops)

### Things to keep
* Content of configuration as it is

### Things to throw out
* ByteValue, DurationUtil classes
* ?Environment class?

### Constraints
* no dependency injection for now
* configuration is only read, never written (might be different if configuration needs to be migrated)
* configuration is local to the broker 

### Nice to Haves
* Command line tool to verify and/or migrate configuration
* Write out effective configuration (potentially aggregated from many sources) for reference and debugging purposes

## Research and Evaluation
Candidates (in order of suitability):
* Spring Boot Configuration
* Typesage Config
* Cfg4J
* Apache Commons Configuration2
* MicroProfile Configuration

Discarded Options:
* Owner (only properties files)
* PureConfig (mainly target at Scala)

### Spring Boot Configuration
tldr: Works like a charm

Missing features: none

#### Pro
* Integrates with Java Bean Validation
* Support for Profiles
* Common framework
* Supports Placeholders in configuration (useful for aliasing)
* Lenient mapping
  - env variable: ZEEBEE_BROKER_NETWORK_PORT
  - proprerty file: zeebee.broker.network.port
  - Several different schemes support (kebab case, camel case , all caps)
* Different ways to define boolean values
* There was something clever about when you print out all settings, it just won't dump all environment variables there are

#### Contra
* Requires application to be a Spring Boot application
* Many dependencies
  - 1s startup time for hello world example
  - 8,5 MB jar size for hello world example
* Cannot access nth Element in List from env properties (probably because the other config hasn't been parsed yet)
* didn't complain about missing configuration setting 

### Typesafe Config
tldr: Somewhat stubborn to work with; unfortunately no YAML

Missing features: YAML Support

#### Pro
* Complains about missing configuration settings - automatic basic validation
* Supports Placeholders in configuration (useful for aliasing)
* Good debug options
* Different ways to define boolean values

#### Contra
* Not clear how to indicate optional settings in the configuration when using object binding
* Cannot set "null" explicitly in configuration
* Env variable override is not enabled by default
* Cannot access nth Element in List from env properties
* Maybe too strict for end users. E.g. it complains if a configuration entry is a string instead a List<String>; Spring Boot on the other hand simply created a list with one element
* Error messages in configuration validation not easy to understand

### Cfg4J
tldr: Nice try, not quite what we are looking for 

Missing features:
* Nested Object Binding
* Binding for list of PoJos
* Binding for Duration (not out of the box)
* Binding for maps

#### Pro
* Complains about missing configuration setting

#### Contra
* Configuration classes must be interfaces for binding; these classes must not follow bean conventions
* Small amount of dependencies (500K)

### Apache Commons Configuration
tldr: only low level API

Missing features: no object binding at all

#### Pro
* Very flexible to set up overlay hierarchies
* Supports Placeholders

#### Contra
* Needs additional dependencies, not declared in POM
  - SnakeYaml
  - BeanUtils

### MicroProfile Config
tldr: Bad first impression; maybe unfarily discarded too early

This evaluation was rather short. It was discarded quickly for the following reasons:
* No support for YAML within the standard
* Binding only works inside a full fledged CDI container
* Overlay structure of different configuration sources pretty much hardwired and difficult to overwrite

### Shortlist
* Spring Boot Configuration
* Typesafe configuation

## Decisions
* Proceed with preparing a prototype based on Spring Boot Configuration
* This should then be tested in the cloud environment to assess any possible side effects or negative impact on performance
* Store default configuration options in the code (not in a file inside the jar)
  - This makes it easier for tests, because they can simply instantiate the configuration object and overwrite what they need - instead of starting with a blank configuration and setting lots of options that are irrelevant to the test)

## Design

* Exporters should become a map instead of a list to be able to better address individual elements

### Preparation
* Move and rename tests #3836
* Remove references to 'maxMessageCount' #3837
* Remove reference to 'contactPoint' #3838
* Remove ByteValue #3841
* Remove DurationUtil #3842
* Add configuration classes for backpressure configuration #3843

### Realization
* Implement prototype for Spring Boot making Broker a Spring Boot application #3862 
* Implement prototype for Spring Boot without making Broker a Spring Boot application #3863
* Use Spring Boot to read configuration #3869
* Document configuration options #3968
* Remove obsolete code #3969
